Safeguard discard

isc-hwojnick opened the pull request:
fixes #455

isc-hwojnick committed:
mostly done with changes

isc-hwojnick committed:
feat: discard safeguard finished

isc-hwojnick committed:
docs: update changelog

isc-tleavitt commented on cls/SourceControl/Git/Utils.cls:
```diff
@@ -1917,6 +1917,7 @@ ClassMethod SyncIrisWithRepoThroughDiff(ByRef files, ByRef filterToFiles, invert
     set addedFiles = $extract(addedFiles, 2, *)
     
     if (deletedFiles '= ""){
+        set ^mtempdeletedFiles($i(^mtempdeletedFiles)) = deletedFiles 
```
Remove debug code

isc-tleavitt reviewed (changes requested):
This also needs to apply to the menu-based discard (ClassMethod Revert in SourceControl.Git.Utils) which uses the checkout command rather than restore.

isc-etamarch commented:
> This also needs to apply to the menu-based discard (ClassMethod Revert in SourceControl.Git.Utils) which uses the checkout command rather than restore.

What do you mean by menu-based discard? Where would you access that in the UI?

isc-etamarch committed:
Fix discarding external files (not in IRIS)

isc-etamarch committed:
Merge branch 'main' into safeguard-discard

isc-etamarch committed:
Replace removed discarded-state code

isc-etamarch committed:
Fix bug in webdriver handling

isc-etamarch committed:
Fix SQL issues, fix file restoration

isc-etamarch committed:
fix macro call

Replaced "$$$THROWONERROR" with "$$$ThrowOnError"

isc-etamarch committed:
Fix file restore

Fix file restore for files whose directories have been deleted. Refresh uncommited queue after file discard

isc-etamarch committed:
Safeguard-discard on menu-based discard

isc-etamarch committed:
Fix UI and post-discard refresh

Fixed the UI so that the "Restore" button in the discarded files is always visible
Run a refresh of the uncommitted queue after the discard

isc-etamarch committed:
Fix restore files after discard

isc-etamarch committed:
Remove from uncommitted queue after discard

isc-etamarch committed:
Fix discarded items still in uncommitted queue

isc-etamarch committed:
typo

isc-etamarch committed:
Bump version

isc-etamarch committed:
Merge branch 'main' into safeguard-discard

isc-tleavitt commented on cls/SourceControl/Git/DiscardState.cls:
```diff
@@ -0,0 +1,138 @@
+Class SourceControl.Git.DiscardState Extends (%Persistent, %JSON.Adaptor)
```
An issue I'm just noticing here: this won't play well with package manager-based editing of multiple packages with distinct repo roots in the same namespace. I'd be happy to say we can't do that just yet and file another issue to pick it up in the future (not necessarily for 2.6.0).

isc-tleavitt commented on cls/SourceControl/Git/DiscardState.cls:
```diff
@@ -0,0 +1,138 @@
+        
+        if (InternalName = "") {
+            // If not in IRIS
+            set externalName = ##class(%File).Construct($Get(^SYS("SourceControl","Git","settings","namespaceTemp")),name)
```
Can we replace the direct global reference with a method call to `SourceControl.Git.Utils`?

isc-tleavitt commented on cls/SourceControl/Git/DiscardState.cls:
```diff
@@ -0,0 +1,138 @@
+{
+    set currentBranch = ##class(SourceControl.Git.Utils).GetCurrentBranch()
+
+    set query = "SELECT ID FROM SourceControl_Git.DiscardState WHERE branch = ?"
```
TODO: filter by repo and branch (in future work, for IPM awareness)

isc-etamarch replied:
Where do we store the repo name?

isc-tleavitt replied:
Maybe it would be better to say "context and branch" where context is a namespace or IPM package name.

isc-tleavitt replied:
But again, should spin out a separate issue for this.

isc-tleavitt commented on cls/SourceControl/Git/DiscardState.cls:
```diff
@@ -0,0 +1,138 @@
+
+Property Username As %String [ Required ];
+
+Property Branch As %String [ Required ];
```
Bitmap index this

isc-tleavitt commented on cls/SourceControl/Git/DiscardState.cls:
```diff
@@ -0,0 +1,138 @@
+
+Property Name As %String [ Required ];
+
+Property Contents As %Stream.GlobalCharacter(LOCATION = "^DiscardStateContents");
```
Let's put this in a better named global e.g. `^SourceControl.Git.DiscardS`

isc-tleavitt commented on cls/SourceControl/Git/WebUIDriver.cls:
```diff
@@ -137,6 +137,10 @@ ClassMethod HandleRequest(pagePath As %String, InternalName As %String = "", Out
                 set handled = 1
+            } elseif (pathStart = "discarded-states") {
+                set discardsInBranch = ##class(SourceControl.Git.DiscardState).DiscardStatesInBranch()
+                do %data.Write(discardsInBranch.%ToJSON())
```
Should do discardsInBranch.%ToJSON(%data) instead

isc-tleavitt commented on git-webui/src/share/git-webui/webui/js/git-webui.js:
```diff
@@ -1937,6 +1947,112 @@ webui.DiffView = function(sideBySide, hunkSelectionAllowed, parent, stashedCommi
+        $('.external-name').text('');
+        
+
+        // TODO: Empty the discarded list properly
```
Is this still relevant?

isc-etamarch committed:
Fix issues

isc-etamarch committed:
Merge branch 'main' into safeguard-discard

isc-tleavitt reviewed (approved):
I think this looks fine from a code standpoint but want to kick the tires on it before merging. @isc-pbarton could you review too?

isc-pbarton commented on git-webui/release/share/git-webui/webui/js/git-webui.js:
```diff
@@ -1937,6 +1947,108 @@ webui.DiffView = function(sideBySide, hunkSelectionAllowed, parent, stashedCommi
+                    });
+                });
+
+                $('.file-contents').html('<pre>' + discardedState.Contents + '</pre>');
```
We need to escape HTML tags in discardedState.Contents. As is, XData blocks like a production definition will not display correctly:
![image](https://github.com/user-attachments/assets/751f6fc2-b42b-40b3-b13c-7aee09869ea8)

isc-pbarton commented on cls/SourceControl/Git/DiscardState.cls:
```diff
@@ -0,0 +1,140 @@
+
+    // Add file to source-control / IRIS
+    if '..ExternalFile {
+        do ##class(SourceControl.Git.Utils).ImportItem(..Name, 1, 1)
```
We should compile after importing. This is especially important for productions because if the class isn't compiled, we won't see anything changed in the interoperability portal.

isc-etamarch committed:
Merge branch 'main' into safeguard-discard

isc-etamarch committed:
Compile discarded classes and fix file contents

Replaced HTML characters with escaped versions, made sure discarded files are compiled after import
